src/wavelet_matrix.rs

```diff
@@ -0,0 +1,559 @@
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// assert_eq!(wm.rank_range(0..22, 'o' as usize), 4);
+    /// ```
+    pub fn rank_range(&self, range: Range<usize>, val: usize) -> usize {
+        let mut start_pos = range.start;
+        let mut end_pos = range.end;
+        for depth in 0..self.bit_length as usize {
+            let bit = Self::get_msb(val, depth, self.bit_length);
+            let rsbv = &self.layers[depth];
+            if bit {
+                start_pos = rsbv.rank1(start_pos) + rsbv.num_zeros();
+                end_pos = rsbv.rank1(end_pos) + rsbv.num_zeros();
+            } else {
+                start_pos = rsbv.rank0(start_pos);
+                end_pos = rsbv.rank0(end_pos);
+            }
+        }
+        (start_pos..end_pos).len()
+    }
+
+    /// Gets the occurrence position of `rank`-th `val` in [0, n).
+    ///
+    /// # Arguments
+    ///
+    /// - `rank`: Rank to be searched.
+    /// - `val`: Integer to be searched.
+    ///
+    /// # Complexity
+    ///
+    /// - $`O(\log \sigma)`$
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+    ///
+    /// let mut wmb = WaveletMatrixBuilder::new();
+    /// let s = "tobeornottobethatisthequestion";
+    /// s.chars().for_each(|c| wmb.push(c as usize));
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// assert_eq!(wm.select(2, 't' as usize), 9);
+    /// ```
+    pub fn select(&self, rank: usize, val: usize) -> usize {
+        self.select_helper(rank, val, 0, 0)
+    }
+
+    fn select_helper(&self, rank: usize, val: usize, pos: usize, depth: usize) -> usize {
+        if depth == self.bit_length {
+            return pos + rank;
+        }
+        let mut pos = pos;
+        let mut rank = rank;
+        let bit = Self::get_msb(val, depth, self.bit_length);
+        let rsbv = &self.layers[depth];
+        if bit {
+            let zeros = rsbv.num_zeros();
+            pos = rsbv.rank1(pos) + zeros;
+            rank = self.select_helper(rank, val, pos, depth + 1);
+            rsbv.select1(rank - zeros)
+        } else {
+            pos = rsbv.rank0(pos);
+            rank = self.select_helper(rank, val, pos, depth + 1);
+            rsbv.select0(rank)
+        }
+    }
+
+    /// Gets `k+1`-th smallest value in the given `range`.
+    ///
+    /// # Arguments
+    ///
+    /// - `range`: Position to be searched.
+    /// - `k`: Integer to be searched.
+    ///
+    /// # Complexity
+    ///
+    /// - $`O(\log \sigma)`$
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
+    ///
+    /// let mut wmb = WaveletMatrixBuilder::new();
+    /// let s = "tobeornottobethatisthequestion";
+    /// s.chars().for_each(|c| wmb.push(c as usize));
+    ///
+    /// let wm = wmb.build().unwrap();
+    /// assert_eq!(wm.quantile(0..3, 0), 'b' as usize); // min in "tob" should be "b"
+    /// ```
```

Comment:
```suggestion
    /// Gets `k`-th smallest value in the given `range`.
    ///
    /// # Arguments
    ///
    /// - `range`: Position range to be searched.
    /// - `k`: Integer to be searched (`k >= 0`).
    ///
    /// # Complexity
    ///
    /// - $`O(\log \sigma)`$
    ///
    /// # Examples
    ///
    /// ```
    /// use sucds::{WaveletMatrix, WaveletMatrixBuilder};
    ///
    /// let mut wmb = WaveletMatrixBuilder::new();
    /// let s = "tobeornottobethatisthequestion";
    /// s.chars().for_each(|c| wmb.push(c as usize));
    ///
    /// let wm = wmb.build().unwrap();
    /// assert_eq!(wm.quantile(0..3, 0), 'b' as usize); // The zero-th in "tob" should be "b"
    /// ```
```

In this repository, k is unified as starting from 0 (such as other selection data structures).